### PR TITLE
Fix container-template calling template-functions with yaml marshalling instead of json

### PR DESCRIPTION
**Pull request by vhvb1989**
Initially, for Aspire, we used to call `json.Marshal(v)` to set the env values for a container template ONLY when: 

```go
!strings.HasPrefix(v, "{") || !strings.HasSuffix(v, "}")
```

On PR: https://github.com/Azure/azure-dev/pull/3095, we changed this to ALWAYS call `json.Marshal(v)` to ensure using strings (see full explanation on PR).

The change addressed a requirement from Aspire Prev2 where fields are expected as strings, but it broke the connection string resolution from the template, as one expression for the go-template like:

```go
value: {{ function "arg" }}
```

Was now written to the yaml as:

```go
value: \"{{ function \\\"arg\\\" }}\"
```

This breaks the template execution, which is not expecting to find the escape `\` for the arg.

This PR changes from using json marshalling to yaml. With yaml, the value is written as:

```go
value: '{{ function "arg" }}'
```

Fields like `true` will continue to be saved as a string, as well as complex expressions (combines replacement with static code). But the quotes for function arguments in the templates won't be affected.

fix: https://github.com/Azure/azure-dev/issues/3126

### CODE REVIEW

**Comment by davidfowl** [comment]
We need a regression test.

File: cli/azd/pkg/apphost/generate.go [hunk 1]
```diff
@@ -828,9 +828,11 @@ func (b *infraGenerator) buildEnvBlock(env map[string]string, manifestCtx *genCo
 		// And YAML rules would treat the above as the value being a boolean instead of a string, which the container
 		// app service expects.
 		//
-		// JSON marshalling the string value will give us something like `"true"` (with the quotes, and any escaping
+		// YAML marshalling the string value will give us something like `"true"` (with the quotes, and any escaping
 		// that needs to be done), which is what we want here.
-		jsonStr, err := json.Marshal(res)
+		// Do not use JSON marshall as it would escape the quotes within the string, breaking the meaning of the value.
+		// yaml marshall will use 'some text "quoted" more text' as a valid yaml string.
+		jsonStr, err := yaml.Marshal(res)
```
Comment (weikanglim) [inline cli/azd/pkg/apphost/generate.go:835]: I'm trying to wrap my head around what this ends up being. `yaml.Marshal` generally would produce marshal the input into a YAML document. `json.Marshal` produces the same, but a single node is a valid JSON document..

What does `yaml.Marshal([]byte("true"))` produce?

Comment (weikanglim) [reply cli/azd/pkg/apphost/generate.go:835]: Ah, looks like `yaml.Marshal` accepts `string` as well, `yaml.Marshal("true")` returns the literal value `"true"`

Comment (weikanglim) [reply cli/azd/pkg/apphost/generate.go:835]: `yaml.Marshal` marshals the string-value as a YAML string, so this works as expected.

**Comment by ellismg** [comment]
@vhvb1989 - I pushed a commit that leveraged the baseline stuff I added for testing aspire manifest generation to add a regression test.

This also changed how some strings were emitted in some of the baselines, so I accepted the updates there.

Thanks for digging into this and fixing it.  Hope I saved you some time with the test.

File: cli/azd/pkg/apphost/generate.go [hunk 2]
```diff
@@ -828,14 +828,16 @@ func (b *infraGenerator) buildEnvBlock(env map[string]string, manifestCtx *genCo
 		// And YAML rules would treat the above as the value being a boolean instead of a string, which the container
 		// app service expects.
 		//
-		// JSON marshalling the string value will give us something like `"true"` (with the quotes, and any escaping
-		// that needs to be done), which is what we want here.
-		jsonStr, err := json.Marshal(res)
+		// YAML marshalling the string value will give us something like `"true"` (with the quotes, and any escaping
+		// that needs to be done), which is what we want here. The TrimSpace is to remove any leading or trailing whitespace
+		// in the string, as yaml.Marshal("true") will give back a string with a trailing newline.
+		yamlStr, err := yaml.Marshal(res)
```
Comment (ellismg) [inline cli/azd/pkg/apphost/generate.go:834]: I do wonder if we should remove the responsibility of this yaml formatting out of this layer and into the templates.  I know in helm (which also uses text templates to produce yaml manifests) there are helpers that you can call in the go text template to get yaml values like this.  That might make things easier long term.  Something for us to think about longer term, I think this is the most direct fix we can make right now.

**Review by ellismg — APPROVED** [review]
I'm happy with this, but I gained lot of confidence in the change by writing the test so someone else should check my work.

We'll want to take this for the preview 2 release as otherwise we'll have a regression.

File: cli/azd/pkg/apphost/generate.go (hunk 2 above)
Comment (ellismg) [reply cli/azd/pkg/apphost/generate.go:833]: (feel free to resolve this when you merge this, @vhvb1989, no action needs to be taken, just wanted to give the thought in some context).

**Review by weikanglim — APPROVED** [review]

**Pull request merged**